lib/src/iterators.rs

```diff
@@ -0,0 +1,81 @@
+use serialize::hex::FromHex;
+
+use std::fs;
+use std::fs::ReadDir;
+use std::io::Result;
+use std::path::Path;
+
+pub struct StoredChunks {
+    dir: ReadDir,
+    next_level: Option<Box<StoredChunks>>,
+    digest_size: usize,
+}
+
+impl StoredChunks {
+    pub fn new(root: &Path, digest_size: usize) -> Result<StoredChunks> {
+        Ok(StoredChunks {
+            dir: try!(fs::read_dir(root)),
+            next_level: None,
+            digest_size: digest_size,
+        })
+    }
+}
+
+impl Iterator for StoredChunks {
+    type Item = Result<Vec<u8>>;
+
+    fn next(&mut self) -> Option<Result<Vec<u8>>> {
+        // Check if we have an iterator for the next level, if so we process from that iterator.
+        let is_some = self.next_level.is_some();
```

Comment:
This piece looks suspicious. Are you sure `if let Some(ref mut next_level) = self.next_level.as_mut() {` wouldn't make it simpler?